### PR TITLE
Optimization: convert vector slice indices to floats

### PR DESCRIPTION
**Pull request by kcpevey**
# Description

Numpy array slicing with object dtypes is much slower than defined (float) slicing. Defining the dtype here gives us a major performance boost. 

This bit of code is nearly identical to `Points` where this performance improvement was applied in [this PR](https://github.com/napari/napari/pull/4648)

## Type of change
performance optimization

# References

With the 2D benchmark, we don't see significant improvements (though it is no worse, either), but with the 3D benchmark, there are huge improvements, e.g. `175±4ms` vs `3.46±0.03ms`. 

Results on `main`:
```
(napari_bench) benchmark$ asv run --python=same --bench Vectors2DSuite.time_set_view_slice
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_opt_miniconda3_envs_napari_bench_bin_python
[ 50.00%] ··· Running (benchmark_vectors_layer.Vectors2DSuite.time_set_view_slice--).
[100.00%] ··· benchmark_vectors_layer.Vectors2DSuite.time_set_view_slice                                                                                                                                                                                   ok
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16     19.5±0.04μs 
                 64      20.0±0.1μs 
                256      21.3±0.2μs 
                1024     24.3±0.4μs 
                4096    32.4±0.09μs 
               16384     64.5±0.8μs 
               65536      190±1μs   
              ======== =============

(napari_bench) benchmark$ asv run --python=same --bench Vectors3DSuite.time_set_view_slice
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_opt_miniconda3_envs_napari_bench_bin_python
[ 50.00%] ··· Running (benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice--).
[100.00%] ··· benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice                                                                                                                                                                                   ok
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16       570±5μs   
                 64       702±6μs   
                256       1.22±0ms  
                1024    3.29±0.05ms 
                4096     11.6±0.2ms 
               16384     44.2±0.2ms 
               65536      175±4ms   
              ======== =============
```

Results on this branch:
```
(napari_bench) benchmark$ asv run --python=same --bench Vectors2DSuite.time_set_view_slice
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_opt_miniconda3_envs_napari_bench_bin_python
[ 50.00%] ··· Running (benchmark_vectors_layer.Vectors2DSuite.time_set_view_slice--).
[100.00%] ··· benchmark_vectors_layer.Vectors2DSuite.time_set_view_slice                                                                                                                                                                                   ok
[100.00%] ··· ======== ============
               param1              
              -------- ------------
                 16     19.4±0.1μs 
                 64     20.3±0.2μs 
                256     20.8±0.1μs 
                1024    23.5±0.1μs 
                4096    31.8±0.1μs 
               16384    63.6±0.2μs 
               65536    189±0.5μs  
              ======== ============

(napari_bench) benchmark$ asv run --python=same --bench Vectors3DSuite.time_set_view_slice
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_opt_miniconda3_envs_napari_bench_bin_python
[ 50.00%] ··· Running (benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice--).
[100.00%] ··· benchmark_vectors_layer.Vectors3DSuite.time_set_view_slice                                                                                                                                                                                   ok
[100.00%] ··· ======== =============
               param1               
              -------- -------------
                 16       517±6μs   
                 64       532±6μs   
                256       533±3μs   
                1024      573±8μs   
                4096      713±7μs   
               16384    1.30±0.03ms 
               65536    3.46±0.03ms 
              ======== =============
```

# How has this been tested?

- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/layers/vectors/vectors.py
Comment (Czaki) [inline napari/layers/vectors/vectors.py]: if this variable is used only in the `if` branch then I prefer to move its definition also in this if branch.

File: napari/layers/vectors/vectors.py
Comment (andy-sweet) [inline napari/layers/vectors/vectors.py]: This is identical to the comment in `Points._slice_data`, except the bit about `Points._round_index` which is missing for `Vectors`, so slice indices are currently rounded. That means that casting to `float` here is in general lossy, but only for really large integers.

We might consider using the following form instead to allow numpy to decide what dtype to use (i.e. `int` vs. `float`): https://github.com/napari/napari/pull/4648/files#r893051191

Alternatively, maybe `Vectors` layers shouldn't be casting slice indices to ints for the same reason that `Points` layers shouldn't.

File: napari/layers/vectors/vectors.py
Comment (andy-sweet) [inline napari/layers/vectors/vectors.py:645]: Though the structure is a little different, this code block is almost identical to the one in `Points._slice_data`. The only major difference is here we calculate `projected_lengths` instead of `sizes`, but from there it looks like the logic is the same once you swap `alpha_match` for `size_match`.

Ideally, I'd like this method to call the same function. That way we don't need to keep the code and comments in sync.

**Review by andy-sweet — COMMENTED** [review]
Benchmark results look great! For that reason, I'd probably be OK with this as is. But I still have a refactor itch around making `Points._slice_data` and `Vectors._slice_data` share more code and comments when they're doing very similar things. Feel free to push back on that if you think that's too much though.

File: napari/layers/vectors/vectors.py
Comment (brisvag) [reply napari/layers/vectors/vectors.py:645]: Agree! On the other hand, I think it could definitely be done in a separate PR as to not block this improvement.

Comment (kcpevey) [reply napari/layers/vectors/vectors.py:645]: I briefly looked at this, but there are some bits which aren't 1:1 that will take some thought to work out. It could be done, but I think leaving it to a separate PR is best.

File: napari/layers/vectors/vectors.py
Comment (andy-sweet) [inline napari/layers/vectors/vectors.py]:
```suggestion
        # promote slicing plane to array so we can index into it, project as type float
```

**Review by andy-sweet — APPROVED** [review]
Looks good to me without the larger refactor. Maybe @brisvag can also quickly review and confirm?

File: napari/layers/vectors/vectors.py
Comment (brisvag) [inline napari/layers/vectors/vectors.py]: I guess all this can go inside the `if` clause right? Not a big deal though, so feel free to ignore.

**Review by brisvag — APPROVED** [review]
LGTM, minor comment that can be ignored.

File: napari/layers/vectors/vectors.py
Comment (kcpevey) [reply napari/layers/vectors/vectors.py]: That's true. I've made the change.

**Comment by sofroniewn** [comment]
This is great @kcpevey - love seeing those benchmarks. This is independent of #4846 I think so we can merge.

I'm good to merge now, though maybe @kcpevey do you want to create a new issue with @andy-sweet suggestion here

> But I still have a refactor itch around making Points._slice_data and Vectors._slice_data share more code and comments when they're doing very similar things.

I also like this idea, but it's something that could come in a followup after this merges.

If that sounds good to you @kcpevey let me know and i'll merge today. thanks!!

**Review by sofroniewn — APPROVED** [review]

**Comment by kcpevey** [comment]
@sofroniewn This can be merged whenever you're ready. I've moved @andy-sweet's comment into [an issue](https://github.com/napari/napari/issues/4896)

**Comment by andy-sweet** [comment]
This has been open for a while and has 3 approvals, so I'm merging now.

**Pull request merged**
